File: leetcode/UniquePaths.py
"""
Remember:
1. How to declare 2d array
2. Iterate 2D array
"""
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        
        if m == n == 1:
            return 1;
        
        dp = [[1] * n for i in range(m)]
        
        for row in range(1, len(dp)):
            for col in range(1, len(dp[row])):
                
                dp[row][col] = dp[row][col - 1] + dp[row - 1][col]
                
                
        return dp[m - 1][n - 1]

File: leetcode/UniquePaths.java
/**
 * Remember: 
   1. Initialize the 2d array with 1
   2. start with 1,1
 */
class Solution {
    public int uniquePaths(int m, int n) {
        
        if (m == 1 && n == 1) {
            return 1;
        }
        
        int[][] dp = new int[m][n];
        
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j  < n; j++){
                dp[i][j] = 1;
            }
        }
        
        
        for (int i = 1; i < m; i++) {
            for (int j  = 1; j < n; j++) {
                dp[i][j] = dp[i][j - 1] + dp[i - 1][j];
            }
        }
        
        return dp[m - 1][n - 1];
        
    }
}
